README.md
# csvmod

A stream-oriented CSV modification tool. Like a stripped-down "sed"
command, but for tabular data.

## Installation

``` bash
$ pip install csvsed
```

## Usage and Examples

Installation of the `csvsed` python package also installs the `csvsed`
command-line tool. Use `csvsed --help` for all command line options,
but here are some examples to get you going. Given the input file
``sample.csv``:

```
Employee ID,Age,Wage,Status
8783,47,"104,343,873.83","All good, but nowhere to go."
2003,32,"98,878,784.00",A-OK
```

Removing thousands-separators from the "Wage" column using the "s"
(substitute) modifier:

``` bash
$ cat sample.csv | csvsed -c Wage s/,//g
Employee ID,Age,Wage,Status
8783,47,104343873.83,"All good, but nowhere to go."
2003,32,98878784.00,A-OK
```

Convert/extract some text using the "s" (substitute) and "y"
(transliterate) modifiers:

``` bash
$ cat sample.csv | csvsed -c Status 's/^All (.*),.*/\1/' \
  | csvsed -c Status 's/^A-(.*)/\1/' \
  | csvsed -c Status 'y/a-z/A-Z/'
Employee ID,Age,Wage,Status
8783,47,"104,343,873.83",GOOD
2003,32,"98,878,784.00",OK
```

Square the "Age" column using the "e" (execute) modifier:

``` bash
$ cat sample.csv | csvsed -c Age 'e/xargs -I {} echo "{}^2" | bc/'
Employee ID,Age,Wage,Status
8783,2209,"104,343,873.83","All good, but nowhere to go."
2003,1024,"98,878,784.00",A-OK
```

That, however, called the external program for each column (quite
inefficient with large data sets)... so let's do that more
efficiently, with a "continuous" mode program. Given the following
``id2name.py`` program which takes a CSV on STDIN with a single column
(an employee ID) and writes a CSV to STDOUT with the IDs converted to
names:

``` python
#!/usr/bin/env python
import sys, csvkit
table = {'8783': 'ElfenKyng', '2003': 'Stradivarius'}
# NOTE: *not* using csvkit's reader because it reads-ahead
# causing problems since this must be stream-oriented...
writer = csvkit.CSVKitWriter(sys.stdout)
while True:
  item = sys.stdin.readline()
  if not item: break
  item = item.strip()
  writer.writerow([table[item] if item in table else item])
  sys.stdout.flush()
```

Then the following will efficiently convert the 'Employee ID' column
to names:

``` bash
$ cat sample.csv | csvsed -c 'Employee ID' 'e/id2name.py/c'
Employee ID,Age,Wage,Status
ElfenKyng,47,"104,343,873.83","All good, but nowhere to go."
Stradivarius,32,"98,878,784.00",A-OK
```


CHANGELOG.rst
=========
ChangeLog
=========


v0.2.4
======

* Artificially bumped version to resolve version conflict


v0.2.3
======

* Updated all file preambles to match license


csvsed/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@metagriffin.net>
# date: 2011/04/08
# copy: (C) Copyright 2011-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

import unittest, StringIO, csvkit
from . import sed

#------------------------------------------------------------------------------
def run(source, modifiers, header=True):
  src = StringIO.StringIO(source)
  dst = StringIO.StringIO()
  reader = csvkit.CSVKitReader(src)
  reader = sed.CsvFilter(reader, modifiers, header=header)
  writer = csvkit.CSVKitWriter(dst)
  for row in reader:
    writer.writerow(row)
  return dst.getvalue()

#------------------------------------------------------------------------------
class TestSed(unittest.TestCase):

  baseCsv = '''\
header 1,header 2,header 3,header 4,header 5
field 1.1,field 1.2,field 1.3,field 1.4,field 1.5
field 2.1,field 2.2,field 2.3,field 2.4,field 2.5
field 3.1,field 3.2,field 3.3,field 3.4,field 3.5
'''

  #----------------------------------------------------------------------------
  def test_charRanges(self):
    self.assertEqual('abcdef', sed.cranges('a-f'))
    self.assertEqual('a-f', sed.cranges('a\-f'))
    self.assertEqual('abc-', sed.cranges('abc-'))
    self.assertEqual('-abc', sed.cranges('-abc'))
    self.assertEqual('a\]^_z', sed.cranges('a\\\\-_z'))
    self.assertEqual('abcdefg', sed.cranges('a-c-e-g'))

  #----------------------------------------------------------------------------
  def test_modifier_y_directcall(self):
    self.assertEqual(sed.Y_modifier('y/abc/def/')('b,a,c'), 'e,d,f')
    self.assertEqual(sed.Y_modifier('y/abc/def/')('b,A,C'), 'e,A,C')
    self.assertEqual(sed.Y_modifier('y/abc/def/i')('b,A,C'), 'e,d,f')
    self.assertEqual(sed.Y_modifier('y/a-z/A-Z/')('Back-Up'), 'BACK-UP')
    self.assertEqual(sed.Y_modifier('y/a\-z/A~Z/')('Back-Up'), 'BAck~Up')

  #----------------------------------------------------------------------------
  def test_modifier_y_toupper(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
field 1.1,field 1.2,FIELD 1.3,field 1.4,field 1.5
field 2.1,field 2.2,FIELD 2.3,field 2.4,field 2.5
field 3.1,field 3.2,FIELD 3.3,field 3.4,field 3.5
'''
    self.assertEqual(run(self.baseCsv, {2: 'y/a-z/A-Z/'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_directcall(self):
    self.assertEqual(sed.S_modifier('s/a/b/')('abcabc'), 'bbcabc')
    self.assertEqual(sed.S_modifier('s/a/b/g')('abcabc'), 'bbcbbc')
    self.assertEqual(sed.S_modifier('s/a/b/g')('abcABC'), 'bbcABC')
    self.assertEqual(sed.S_modifier('s/a/b/gi')('abcABC'), 'bbcbBC')

  #----------------------------------------------------------------------------
  def test_modifier_s_noflags(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
xield 1.1,field 1.2,field 1.3,field 1.4,field 1.5
xield 2.1,field 2.2,field 2.3,field 2.4,field 2.5
xield 3.1,field 3.2,field 3.3,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/./x/'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_gflag(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
xxxxxxxxx,field 1.2,field 1.3,field 1.4,field 1.5
xxxxxxxxx,field 2.2,field 2.3,field 2.4,field 2.5
xxxxxxxxx,field 3.2,field 3.3,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/./x/g'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_multicol(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
xxxxxxxxx,field 1.2,yyyyyyyyy,field 1.4,field 1.5
xxxxxxxxx,field 2.2,yyyyyyyyy,field 2.4,field 2.5
xxxxxxxxx,field 3.2,yyyyyyyyy,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/./x/g', 2: 's/./y/g'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_colbyname(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
xxxxxxxxx,field 1.2,yyyyyyyyy,field 1.4,field 1.5
xxxxxxxxx,field 2.2,yyyyyyyyy,field 2.4,field 2.5
xxxxxxxxx,field 3.2,yyyyyyyyy,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(
      run(self.baseCsv, {'header 1': 's/./x/g', 'header 3': 's/./y/g'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_nomatch(self):
    chk = self.baseCsv
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/[IE]/../'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_iflag(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
f..eld 1.1,field 1.2,field 1.3,field 1.4,field 1.5
f..eld 2.1,field 2.2,field 2.3,field 2.4,field 2.5
f..eld 3.1,field 3.2,field 3.3,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/[IE]/../i'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_multiflag(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
f....ld 1.1,field 1.2,field 1.3,field 1.4,field 1.5
f....ld 2.1,field 2.2,field 2.3,field 2.4,field 2.5
f....ld 3.1,field 3.2,field 3.3,field 3.4,field 3.5
'''
    self.assertMultiLineEqual(run(self.baseCsv, {0: 's/[IE]/../ig'}), chk)

  #----------------------------------------------------------------------------
  def test_modifier_s_remove(self):
    src = 'cell 1,"123,456,789.0"\n'
    chk = 'cell 1,123456789.0\n'
    self.assertMultiLineEqual(run(src, {1: 's/,//g'}, header=False), chk)

  #----------------------------------------------------------------------------
  def test_modifier_e_directcall(self):
    self.assertEqual(sed.E_modifier('e/tr ab xy/')('b,a,c'), 'y,x,c')
    self.assertEqual(sed.E_modifier('e/xargs -I {} echo "{}^2" | bc/')('4'), '16')

  #----------------------------------------------------------------------------
  def test_modifier_e_multipipe(self):
    chk = '''\
header 1,header 2,header 3,header 4,header 5
field 1.1,field 1.2,field 1.3,1.96,field 1.5
field 2.1,field 2.2,field 2.3,5.76,field 2.5
field 3.1,field 3.2,field 3.3,11.56,field 3.5
'''
    self.assertMultiLineEqual(
      run(self.baseCsv, {3: 'e/cut -f2 -d" " | xargs -I {} echo "scale=3;{}^2" | bc/'}), chk)

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


csvsed/cli.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@metagriffin.net>
# date: 2009/08/04
# copy: (C) Copyright 2009-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

'''
Command-line interface to `csvsed.sed`.
'''

import re, sys

from csvkit import CSVKitReader, CSVKitWriter
from csvkit.cli import CSVKitUtility, CSVFileType, parse_column_identifiers
from csvsed import sed

#------------------------------------------------------------------------------
class CsvSed(CSVKitUtility):

  description = 'A stream-oriented CSV modification tool. Like a ' \
      ' stripped-down "sed" command, but for tabular data.'
  override_flags = 'f'

  #----------------------------------------------------------------------------
  def add_arguments(self):
    self.argparser.add_argument(
      '-c', '--columns',
      dest='columns',
      help='A comma separated list of column indices or names to be modified.')
    # todo: support in-place file modification
    # todo: make sure that it supports backup spec, eg '-i.orig'
    # self.argparser.add_argument(
    #   '-i', '--in-place',
    #   dest='inplace',
    #   help='Modify a file in-place (with a value, specifies that the original'
    #   ' should be renamed first, e.g. "-i.orig")')
    self.argparser.add_argument(
      'expr', metavar='EXPR',
      help='The "sed" expression to evaluate: currently supports substitution'
      ' (s/REGEX/EXPR/FLAGS) and transliteration (y/SRC/DEST/FLAGS).')
    self.argparser.add_argument(
      'file', metavar='FILE',
      nargs='?', type=CSVFileType(), default=sys.stdin,
      help='The CSV file to operate on. If omitted or "-", will read from STDIN.')

  #----------------------------------------------------------------------------
  def main(self):
    reader = CSVKitReader(self.args.file, **self.reader_kwargs)
    cnames = reader.next()
    cids   = parse_column_identifiers(self.args.columns, cnames, self.args.zero_based)
    mods   = {idx: self.args.expr for idx in cids}
    output = CSVKitWriter(self.output_file, **self.writer_kwargs)
    reader = sed.CsvFilter(reader, mods, header=False)
    output.writerow(cnames)
    for row in reader:
      output.writerow(row)

#------------------------------------------------------------------------------
def main():
  utility = CsvSed()
  return utility.main()

if __name__ == '__main__':
  sys.exit(main())

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


csvsed/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@metagriffin.net>
# date: 2009/08/04
# copy: (C) Copyright 2009-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

from sed import *

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


csvsed/sed.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@metagriffin.net>
# date: 2009/08/04
# copy: (C) Copyright 2009-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

'''
A stream-oriented CSV modification tool. Like a stripped-down "sed"
command, but for tabular data.
'''

import re, string, types, subprocess, csvkit, csv
from csvkit.exceptions import ColumnIdentifierError

#------------------------------------------------------------------------------
class InvalidModifierSpec(Exception): pass

#------------------------------------------------------------------------------
class CsvFilter(object):
  def __init__(self, reader, modifiers, header=True):
    '''
    On-the-fly modifies CSV records coming from a csvkit reader object.

    :Parameters:

    reader : iter

      The CSV record source - must support the `next()` call, which
      should return a list of values.

    modifiers : { list, dict }

      Specifies a set of modifiers to apply to the `reader`, which can
      be either a sequence or dictionary of a modifiers to apply. If
      it is a sequence, then the modifiers are applied to the
      equivalently positioned cells in the input records. If it is a
      dictionary, the keys can be integers (column position) or
      strings (column name). In all cases, the modifiers can be one of
      the following:

      * function : takes a single string argument and returns a string
      * string : a sed-like modification specification.

      Currently supported modification specifications:

      * Substitution: "s/REGEX/REPL/FLAGS"

        Replaces regular expression `REGEX` with replacement string
        `REPL`, which can use back references. Supports the following
        flags:

        * i: case-insensitive
        * g: global replacement (otherwise only the first is replaced)
        * l: uses locale-dependent character classes
        * m: enables multiline matching for "^" and "$"
        * s: "." also matches the newline character
        * u: enables unicode escape sequences
        * x: `REGEX` uses verbose descriptors & comments

      * Transliteration: "y/SRC/DST/FLAGS"

        (This is a slightly modified version of sed's "y" command.)

        Each character in `SRC` is replaced with the corresponding
        character in `DST`. The dash character ("-") indicates a range
        of characters (e.g. "a-z" for all alphabetic characters).  If
        the dash is needed literally, then it must be the first or
        last character, or escaped with "\". The "\" character escapes
        itself. Only the "i" flag, indicating case-insensitive
        matching of `SRC`, is supported.

      Note that the "/" character can be any character as long as it
      is used consistently and not used within the specification,
      e.g. ``s|a|b|`` is equivalent to ``s/a/b/``.

    header : bool, optional, default: true

      If truthy (the default), then the first row will not be modified.
    '''
    self.reader    = reader
    self.header    = header
    self.cnames    = None if not header else reader.next()
    self.modifiers = standardize_modifiers(self.cnames, modifiers)

  #----------------------------------------------------------------------------
  def __iter__(self):
    return self

  #----------------------------------------------------------------------------
  def next(self):
    if self.header:
      self.header = False
      return self.cnames
    row = self.reader.next()
    for col, mod in self.modifiers.items():
      row[col] = mod(row[col])
    return row

#------------------------------------------------------------------------------
def standardize_modifiers(cnames, modifiers):
  # TODO: csvkit.grep.standardize_patterns could be refactored to support
  #       this process here as well...
  try:
    # Test to see if dictionary of modifiers
    modifiers = {k: v for k, v in modifiers.items() if v}
  except AttributeError:
    # Fallback to sequence of modifiers
    return {i: spec2modifier(v) for i, v in enumerate(modifiers) if v}
  modifiers = {k: spec2modifier(v) for k, v in modifiers.items()}
  if not cnames:
    return modifiers
  p2 = {}
  for k in modifiers:
    if k in cnames:
      idx = cnames.index(k)
      if idx in modifiers:
        raise ColumnIdentifierError(
          'Column %s has index %i which already has a pattern.' % (k,idx))
      p2[idx] = modifiers[k]
    else:
      p2[k] = modifiers[k]
  return p2

#------------------------------------------------------------------------------
def spec2modifier(obj):
  # obj is function
  if hasattr(obj, '__call__'):
    return obj
  # obj is a specification string
  return eval(obj[0].upper() + '_modifier')(obj)

#------------------------------------------------------------------------------
class S_modifier(object):
  'The "substitution" modifier ("s/REGEX/REPL/FLAGS").'
  def __init__(self, spec):
    super(S_modifier, self).__init__()
    if not spec or len(spec) < 4 or spec[0] != 's':
      raise InvalidModifierSpec(spec)
    sspec = spec.split(spec[1])
    if len(sspec) != 4:
      raise InvalidModifierSpec(spec)
    flags = 0
    for flag in sspec[3].upper():
      flags |= getattr(re, flag, 0)
    self.regex = re.compile(sspec[1], flags)
    self.repl  = sspec[2]
    self.count = 0 if 'g' in sspec[3].lower() else 1
  def __call__(self, value):
    return self.regex.sub(self.repl, value, count=self.count)

#------------------------------------------------------------------------------
def cranges(spec):
  # todo: there must be a better way...
  ret = ''
  idx = 0
  while idx < len(spec):
    c = spec[idx]
    idx += 1
    if c == '-' and len(ret) > 0 and len(spec) > idx:
      for i in range(ord(ret[-1]) + 1, ord(spec[idx]) + 1):
        ret += chr(i)
      idx += 1
      continue
    if c == '\\' and len(spec) > idx:
      c = spec[idx]
      idx += 1
    ret += c
  return ret

#------------------------------------------------------------------------------
class Y_modifier(object):
  'The "transliterate" modifier ("y/SOURCE/DESTINATION/FLAGS").'
  # todo: the python2 string.maketrans & string.translate functions
  #       only work on non-unicode input and csvkit produces unicode
  #       values... so the current 'y' modifier does not use them.
  #       *HOWEVER*, python3's version *does* work, so in py3 mode,
  #       use that!
  def __init__(self, spec):
    super(Y_modifier, self).__init__()
    if not spec or len(spec) < 4 or spec[0] != 'y':
      raise InvalidModifierSpec(spec)
    yspec = spec.split(spec[1])
    if len(yspec) != 4:
      raise InvalidModifierSpec(spec)
    yspec[1] = cranges(yspec[1])
    yspec[2] = cranges(yspec[2])
    if 'i' in yspec[3].lower():
      # self.table = string.maketrans(yspec[1].lower() + yspec[1].upper(),
      #                               2 * yspec[2])
      self.src = yspec[1].lower() + yspec[1].upper()
      self.dst = 2 * yspec[2]
    else:
      # self.table = string.maketrans(yspec[1], yspec[2])
      self.src = yspec[1]
      self.dst = yspec[2]
    if len(self.src) != len(self.dst):
      raise InvalidModifierSpec(spec)
  def __call__(self, value):
    # return string.translate(val, self.table)
    # TODO: this could be *much* more efficient...
    ret = ''
    for ch in value:
      idx = self.src.find(ch)
      if idx < 0:
        ret += ch
      else:
        ret += self.dst[idx]
    return ret

#------------------------------------------------------------------------------
class ReadlineIterator(object):
  'An iterator that calls readline() to get its next value.'
  # NOTE: this is a hack to make csv.reader not read-ahead.
  def __init__(self, f): self.f = f
  def __iter__(self): return self
  def next(self):
    line = self.f.readline()
    if not line: raise StopIteration
    return line

#------------------------------------------------------------------------------
class E_modifier(object):
  'The "execute" external program modifier ("e/PROGRAM+OPTIONS/FLAGS").'
  def __init__(self, spec):
    super(E_modifier, self).__init__()
    if not spec or len(spec) < 3 or spec[0] != 'e':
      raise InvalidModifierSpec(spec)
    espec = spec.split(spec[1])
    if len(espec) != 3:
      raise InvalidModifierSpec(spec)
    espec[2] = espec[2].lower()
    self.command = espec[1]
    self.index   = 1 if 'i' in espec[2] else None
    self.csv     = 'c' in espec[2]
    if not self.csv:
      return
    self.proc = subprocess.Popen(
      self.command, shell=True, bufsize=0,
      stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    self.writer = csvkit.CSVKitWriter(self.proc.stdin)
    # note: not using csvkit's reader because there is no easy way of
    # making it not read-ahead (which breaks the "continuous" mode).
    # self.reader = csvkit.CSVKitReader(self.proc.stdout)
    # todo: fix csvkit so that it can be used in non-read-ahead mode.
    self.reader = csv.reader(ReadlineIterator(self.proc.stdout))
  def __call__(self, value):
    if not self.csv:
      return self.execOnce(value)
    self.writer.writerow([value])
    return self.reader.next()[0].decode('utf-8')
  def execOnce(self, value):
    p = subprocess.Popen(
      self.command, shell=True,
      stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    output, errput = p.communicate(value)
    if p.returncode != 0:
      raise Exception('command "%s" failed: %s' % (self.command, errput))
    if output[-1] == '\n':
      output = output[:-1]
    return output

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------
